### PR TITLE
ODC-6783: Allow hiding console perspectives

### PR DESCRIPTION
**Pull request by debsmita1**
Story:
https://issues.redhat.com/browse/ODC-6783

Add dev catalog types customization API based on https://github.com/openshift/enhancements/pull/1205

### CODE REVIEW

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:135]: Is there a way we can describe a perspective so that a user knows exactly what that means? How are they described in the user docs?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:137]: You only need one of these two, we prefer to keep `+optional`
```suggestion
	// +optional
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:183]: This isn't the correct tag, it should be:
```suggestion
	// +kubebuilder:validation:Resource
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:182]: Maybe give a few examples of expected kinds?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:187]: Are verbs required to be lower case? Can you only set a single verb or should this be a verbs list?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:241]: This needs an enum tag
```suggestion
	// +kubebuilder:validation:Enum:=Enabled;Disabled
	State PerspectiveState `json:"state,omitempty"`
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:231]: Delete this line, this tag isn't valid

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:235]: Delete this line, you already have `+optional`

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:239]: Delete this line, you already have `+optional`

**Review by JoelSpeed — COMMENTED** [review]
Please review our guidance on [how to document APIs](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#write-user-readable-documentation-in-godoc) and then review the fields here, it may inspire you to add additional context to a few of the fields. For example, for an enumerated value, what are the valid values. For fields that have patterns, what are the restrictions on that pattern. If a fields is optional, what does it mean for it to be omitted.

Once you've done that let me know and we can review again

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go]: Removed this type and using the existing `ResourceAttributes` type

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:225]: We tend to discourage using external types because they aren't under our control. They may have additional fields that you don't need within your API or, over time, they may add additional fields which suddenly get added to your API.

My advice would be to copy the ResourceAttributes here if you want to use a similar structure. This also allows you to customise the comments (and therefore generated docs) to your specific API use case

Comment (jerolimov) [reply operator/v1/types_console.go:225]: That is a good point, but we should also keep in mind this is a versioned k8s API (k8s.io/api/authorization/v1), and we use `ResourceAttributes` also in `types_console_quick_start.go` for a similar case:

https://github.com/openshift/api/blob/0b39f81154fa42abe26933181141ed4b4e3ad247/console/v1/types_console_quick_start.go#L70-L74

We would also pass the object 1:1 to an RBAC check. So for me, it makes sense to use than the k8sn API which also supports things like subresources etc. which we might not need for our case. But we want to develop this feature in a (generic) way so that customers can apply their own rules (permission checks) if needed. If we create our own type that supports only resource and verb, this will add IMO unnecessary limits.

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:214]: I would expect that the enum value of PerspectiveState.Enabled should be "Enabled".

See https://go.dev/play/p/Xitfcfrdsqj

```suggestion
	Enabled  PerspectiveState = "Enabled"
	Disabled PerspectiveState = "Disabled"
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
Just one small thing, everything else looks good to me.

**Comment by JoelSpeed** [comment]
Please make sure that this reflects the changes recently updated within the enhancement

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:238]: For a discriminated union, members must be pointers (apart from the discriminator)

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:137]: Is this meant to be a list?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:222]: Are there fields present within ResourceAttributes that you do not intend to support straight away?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:220]: What happens when there are no checks?

What happens when this is configured and `missingAccessReview` is also configured?

Since this is a list, should it be plural rather than singular?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:219]: What would happen if I don't set any of the fields within this struct? Should we enforce that at least one is set? 
```
// +kubebuilder:validation:MinItems:=1
```
(I believe that's the correct incantation)

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:235]: No need for omitempty on a required field

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:245]: Is there any validation we can add here? Is there a set of characters or a length limit we can impose?
Where can a user find these values?

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:247]: Is it optional? What does a perspective mean if we don't set its visibility? It won't do anything right?

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go]: If there are no checks, we show the perspective. Added details as per the enhancement doc

updated it to `required` and `missing`

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go:222]: If we do not set any fields then we enable the perspective

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go:248]: currently, we do not have any criteria for id in our javascript code. The id here should match the extension id of the perspective. So leaving this as is.

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go]: updated it to Required

File: operator/v1/types_console.go
Comment (deads2k) [reply operator/v1/types_console.go]: When a value is interpreted by the consumer of the API, then precise control of the schema is a requirement since the specification of additional fields into an existing v1 API (fully allowed), can cause mishandling of fields and defy user expectation.  For instance, imagine we add another field about storage version to the access review.  If you had code reading and interpreting the value, then the semantics of the field you're referencing would be out of date and a user that expects a reject may suddenly get an allow.

When a value is used fully opaquely and passed through to a matching endpoint, a case can be made to use a type that has a stability guarantee equal to or greater than the stability level used in OCP.

Comment (jerolimov) [reply operator/v1/types_console.go]: Agree. We pass it through an endpoint that understands auth-checks. We don't interpret these permission checks on our own. So I would expect we're fine here.

Comment (jerolimov) [reply operator/v1/types_console.go]: I checked the detail now, we send it to k8s `group: authorization.k8s.io/v1`, `kind: SelfSubjectAccessReview` that get this a (single) `ResourceAttributes`.

File: operator/v1/types_console.go
Comment (jerolimov) [reply operator/v1/types_console.go:222]: You can not specify miniItems: 1 here because it should be okay to specific only missing resources.

**Review by jerolimov — APPROVED** [review]

File: operator/v1/types_console.go
Comment (JoelSpeed) [reply operator/v1/types_console.go:222]: Sorry, I meant `MinProperties` on the `PerspectiveAccessReview` struct would require that either Required or Missing is set, it prevents it from being an empty struct, eg https://github.com/openshift/api/blob/48d78630feb3a1df10bfd2e3797a2b96441727e2/machine/v1/types_controlplanemachineset.go#L214

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:228]: I don't think the current description describes what the end user sees, is it more clear with this suggestion? I'm trying to explain it in terms of what happens to the perspective
```suggestion
	// required defines a list of permission checks. The perspective will only be shown when all checks are successful. When omitted, the access review is skipped and the perspective will not be shown unless it is required to do so based on the configuration of the missing access review list.
	// +optional
	Required []authorizationv1.ResourceAttributes `json:"required,omitempty"`
	// missing defines a list of permission checks. The perspective will only be shown when at least one check fails. When omitted, the access review is skipped and the perspective will not be shown unless it is required to do so based on the configuration of the required access review list.
	// +optional
	Missing []authorizationv1.ResourceAttributes `json:"missing,omitempty"`
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:221]: Is this intuitive? Both fields say they only show the field when their checks pass/fail respectively, so I would have thought if there's no check in the required (which would mean it's not shown) and there's no check in the missing (which would mean it's not shown), then the perspective would not be shown?

Two not shows shouldn't make a show should it?

File: operator/v1/types_console.go
Comment (JoelSpeed) [reply operator/v1/types_console.go:248]: Where does a user find out the ID? We should make a note in the documentation to give them a hint where they might find this

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:225]: Don't use omitempty here, it harms discoverability.

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:137]: Don't use `omitempty` here, it harms the discroverability

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:236]: To allow discoverability, we may need to include the `""` as a valid value. Can you remove the omitempty from all of the fields in this API except the AccessReview field below, then take the API for a test run and check the validations work?

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go:223]: Added this validation

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:137]: Am i right in thinking that the ID in this list can be unique?

We should add a list type to this for server side apply

Xref https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy

I suspect this is an atomic list but let's double check

File: operator/v1/types_console.go
Comment (JoelSpeed) [reply operator/v1/types_console.go:237]: Ignore this comment, doesn't apply as this is an item in a list

File: operator/v1/types_console.go
Comment (JoelSpeed) [reply operator/v1/types_console.go:137]: We discussed this, ID must be unique within the list
```
// +listType=map
// +listMapKey=id
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:242]: I think we should be requiring the member here, which means we can drop the `Empty accessReview...` comment

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:236]: This will require that the `accessReview` is configured iff `state: AccessReview`
```suggestion
// +kubebuilder:validation:XValidation:rule="self.state == 'AccessReview' ?  has(self.accessReview) : !has(self.accessReview)",message="accessReview configuration is required when state is AccessReview, and forbidden otherwise"
type PerspectiveVisibility struct {
```

**Comment by JoelSpeed** [comment]
/lgtm

/hold

I just want to double check with @deads2k that he is ok with adding the CEL now, I have tested it manually and it works as expected IMO but we can't test this until the K8s rebase lands

**Comment by deads2k** [comment]
> I just want to double check with @deads2k that he is ok with adding the CEL now, I have tested it manually and it works as expected IMO but we can't test this until the K8s rebase lands

For new APIs, this is ok.  I don't want to convert any pre-4.12 field prior to the kube udpate.  Feel free to release the hold.

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:248]:
```suggestion
type Perspective struct {
```

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:139]: nit: the type should be singular imho, the attribute name and json name should be still plural.

```suggestion
	Perspectives []Perspective `json:"perspectives"`
```

**Review by jerolimov — CHANGES_REQUESTED** [review]

**Comment by jerolimov** [comment]
Thanks for the update

/lgtm

**Comment by debsmita1** [comment]
updated the go doc for perspective id

**Comment by jerolimov** [comment]
I've tested this API via https://github.com/openshift/console-operator/pull/678 on a cluster bot instance and it works fine. Tested this also together with https://github.com/openshift/console/pull/12014

/lgtm
/label qe-approved

**Pull request merged**
